[#4231] Disallow remote groups (4-2-stable)

alanking opened the pull request:
CI tests to run, all the new tests pass at the bench.

korydraughn commented on server/core/include/user_validation_utilities.hpp:
```diff
@@ -0,0 +1,63 @@
+    /// \throws irods::exception If _user_name does not exist
+    ///
+    /// \since 4.2.11
+    auto get_type(RsComm* _comm, const std::string_view _user_name) -> std::string;
```
if `_comm` is expected to be valid, we should make this a reference instead of a pointer.

alanking replied:
Yeps

korydraughn commented on server/api/src/rsGeneralAdmin.cpp:
```diff
@@ -40,6 +44,58 @@ namespace
+    } // throw_if_group_is_changing_to_user_or_user_is_changing_to_group
+
+    auto throw_if_password_is_being_set_on_a_group(
+        RsComm* _comm,
```
reference instead of pointer?

alanking replied:
Done

korydraughn commented on server/api/src/rsGeneralAdmin.cpp:
```diff
@@ -40,6 +44,58 @@ namespace
     } // contains_whitespace
+
+    auto throw_if_group_is_changing_to_user_or_user_is_changing_to_group(
+        RsComm* _comm,
```
reference instead of pointer?

alanking replied:
Done

alanking commented:
Linking irods/irods_client_icommands#237

trel commented on scripts/irods/lib.py:
```diff
@@ -707,3 +707,26 @@ def add_child_resource(parent_resource_name, child_resource_name, user):
+
+def get_user_type(session, username):
+    """Get the type field of the user indicated by username."""
+    gql = "select USER_TYPE where USER_NAME = '{}'"
```
all four of these in this diff could match more than one user?   `alice#tempZone` and `alice#otherZone`?

alanking replied:
A fine point. CI tests caught this as well, so I'll tinker with this a bit more.

alanking replied:
Added the zone names and adjusted tests. Everything is passing now.

trel commented on server/core/include/administration_utilities.hpp:
```diff
@@ -0,0 +1,48 @@
+    /// If _user_name does not contain the zone name using the octothorpe delimiter, the
+    /// _zone_name will be used. Both of these are available here because the GeneralAdmin
+    /// interface is flexible enough to allow the zone name to be passed through the username
+    /// string or through the generalAdninInp struct.
```
Is there a test where the zone name is both *in* the username string *and* the _zone_name?   And if they don't agree... seems the _user_name string wins?

alanking replied:
Looking at the implementation again, we actually aren't using _zone_name at all. And `iadmin` is passing an empty string for that parameter. So I'm just going to ignore it entirely and look for the zone name in the user name.

trel commented on server/core/src/user_validation_utilities.cpp:
```diff
@@ -0,0 +1,76 @@
+        char user[NAME_LEN]{};
+        parseUserName(_user_name.data(), user, nullptr);
+
+        const std::string qstr = fmt::format("SELECT USER_TYPE where USER_NAME = '{}'", user);
```
same as other comment... what happens if there is both a local and a remote alice?

alanking replied:
Added zone name to the query to avoid duplicates.

alanking commented:
Force-pushed because I had done some branching gymnastics to avoid some recent commits which I have now rebased to include and just went ahead and squashed everything.

trel commented:
very good - seems ready for the #s to me

alanking commented:
#'d
